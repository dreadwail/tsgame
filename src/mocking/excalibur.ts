import { Engine, Logger, Scene } from 'excalibur';

export const createMockLogger = (): Logger => {
  const mock = {
    error: jest.fn(),
    warn: jest.fn(),
  };
  return (mock as any) as Logger;
};

export const createMockEngine = (): Engine => {
  const mock = {
    drawWidth: 800,
    drawHeight: 600,
    halfDrawWidth: 400,
    halfDrawHeight: 300,
    add: jest.fn((...args: any[]) => {
      if (args[1] instanceof Scene) {
        const scene = args[1];
        scene.engine = (mock as any) as Engine;
      }
    }),
    start: jest.fn(() => ({
      then: jest.fn(),
    })),
    goToScene: jest.fn(),
  };
  return (mock as any) as Engine;
};
