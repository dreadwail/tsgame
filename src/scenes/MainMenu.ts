import { BaseAlign, Color, Engine, Label, Scene, TextAlign, Vector } from 'excalibur';
import { ILabelArgs } from 'excalibur/dist/Label';

const menuEntryDefaults: ILabelArgs = {
  fontFamily: 'Sigmar One',
  textAlign: TextAlign.Center,
  fontSize: 50,
  color: Color.White,
  baseAlign: BaseAlign.Middle,
};

export default class MainMenu extends Scene {
  public static sceneName = 'mainMenu';

  private menuEntries: Label[];

  constructor(engine: Engine) {
    super(engine);
    this.menuEntries = [
      this.createMenuEntry('New Game'),
      this.createMenuEntry('Continue'),
      this.createMenuEntry('Exit'),
    ];
  }

  public onInitialize(_engine: Engine) {
    this.menuEntries.forEach(entry => {
      this.add(entry);
    });
  }

  public onActivate() {
    this.flowMenuEntries();
  }

  private createMenuEntry = (text: string) => {
    const label = new Label({ ...menuEntryDefaults, text });
    label.setHeight(label.fontSize);
    return label;
  };

  private flowMenuEntries = () => {
    const { halfDrawWidth, halfDrawHeight } = this.engine;
    let drawY = halfDrawHeight;
    this.menuEntries.forEach(entry => {
      entry.pos = new Vector(halfDrawWidth, drawY);
      drawY += entry.fontSize;
    });
  };
}
