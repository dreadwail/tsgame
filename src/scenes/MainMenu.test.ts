import { Actor, Engine, Label } from 'excalibur';

import { createMockEngine } from '../mocking/excalibur';

import MainMenu from './MainMenu';

describe('MainMenu', () => {
  let engine: Engine;
  let mainMenu: MainMenu;

  beforeEach(() => {
    engine = createMockEngine();
    mainMenu = new MainMenu(engine);
    engine.add(MainMenu.sceneName, mainMenu);
  });

  describe('onInitialize', () => {
    beforeEach(() => {
      mainMenu.onInitialize(engine);
    });

    it('initializes the correct labels', () => {
      const labels = mainMenu.actors
        .filter((actor: Actor): actor is Label => actor instanceof Label)
        .map(label => label.text);

      expect(labels).toEqual(['New Game', 'Continue', 'Exit']);
    });
  });

  describe('onActivate', () => {
    beforeEach(() => {
      mainMenu.onInitialize(engine);
      mainMenu.onActivate();
    });

    it('correctly positions the menu entries', () => {
      const labels = mainMenu.actors
        .filter((actor: Actor): actor is Label => actor instanceof Label)
        .map(label => label.pos);

      expect(labels).toMatchSnapshot();
    });
  });
});
