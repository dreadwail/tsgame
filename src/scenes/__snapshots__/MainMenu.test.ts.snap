// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainMenu onActivate correctly positions the menu entries 1`] = `
Array [
  Vector {
    "x": 400,
    "y": 300,
  },
  Vector {
    "x": 400,
    "y": 350,
  },
  Vector {
    "x": 400,
    "y": 400,
  },
]
`;
